libs/cardano-ledger-test/src/Test/Cardano/Ledger/STS.hs

```diff
@@ -0,0 +1,219 @@
+{-# LANGUAGE ViewPatterns #-}
+
+module Test.Cardano.Ledger.STS where
+
+import Control.State.Transition.Extended
+import Data.Default.Class (Default (def))
+import Data.Foldable
+import Lens.Micro ((^.))
+
+import Test.QuickCheck
+
+import Test.Tasty
+import Test.Tasty.QuickCheck
+
+import Cardano.Ledger.Conway
+import Cardano.Ledger.Conway.Core
+import Cardano.Ledger.Conway.Rules
+
+import Cardano.Ledger.Conway.PParams
+
+import Test.Cardano.Ledger.Generic.Proof
+import Test.Cardano.Ledger.Shelley.Utils
+
+import Cardano.Ledger.CertState
+import Test.Cardano.Ledger.Constrained.Ast
+import Test.Cardano.Ledger.Constrained.Preds.CertState
+import Test.Cardano.Ledger.Constrained.Preds.PParams
+import Test.Cardano.Ledger.Constrained.Preds.Universes
+import Test.Cardano.Ledger.Constrained.Rewrite
+import Test.Cardano.Ledger.Constrained.TypeRep
+import Test.Cardano.Ledger.Constrained.Vars
+
+import qualified Data.Map as Map
+import Data.Maybe
+
+import Test.Cardano.Ledger.Constrained.Env
+
+import Test.Cardano.Ledger.Constrained.Classes
+import Test.Cardano.Ledger.Constrained.Monad
+
+import Test.Cardano.Ledger.Constrained.Shrink
+
+import Test.Cardano.Ledger.Alonzo.AlonzoEraGen ()
+import Test.Cardano.Ledger.Alonzo.EraMapping ()
+import Test.Cardano.Ledger.Constrained.Solver
+
+import Cardano.Ledger.Conway.TxCert
+
+------------------------------------------------------------------------
+-- Generator Setup
+------------------------------------------------------------------------
+
+genFromConstraints :: Era era => Proof era -> OrderInfo -> [Pred era] -> RootTarget era s t -> Gen t
+genFromConstraints proof order cs target = do
+  subst <- genSubstFromConstraints proof order cs
+  env <- monadTyped $ substToEnv subst emptyEnv
+  monadTyped (runTarget env target)
+
+genSubstFromConstraints :: Era era => Proof era -> OrderInfo -> [Pred era] -> Gen (Subst era)
+genSubstFromConstraints proof order cs = do
+  (_, graph) <- compileGenWithSubst order emptySubst cs
+  result <- genDependGraph False proof graph
+  case result of
+    Left errs -> error $ unlines errs
+    Right subst -> pure subst
+
+shrinkFromConstraints :: Era era => OrderInfo -> [Pred era] -> RootTarget era t t -> t -> [t]
+shrinkFromConstraints order cs target val = do
+  let env = saturateEnv (getTarget val target emptyEnv) cs
+  graph <- monadTyped $ compile order $ rewrite cs
+  env' <- shrinkEnv graph env
+  monadTyped $ runTarget env' target
+
+-- | Add variables to the environment that are uniquely determined by the constraints.
+saturateEnv :: Era era => Env era -> [Pred era] -> Env era
+saturateEnv env0 preds = go env0 preds
+  where
+    go env [] = env
+    go env (p : ps)
+      | Just (x, v) <- solveUnknown env p = saturateEnv (storeName x v env) preds
+      | otherwise = go env ps
+
+solveUnknown :: forall era. Era era => Env era -> Pred era -> Maybe (Name era, Payload era)
+solveUnknown env p = case p of
+  SumsTo _ (Var x@(V _ rep acc)) EQL sums
```

Comment:
lots of missing cases when we pattern match against 'p'.  Do we have a plan to cover these as well?

Reply:
Yes, my plan is to implement these as they start to appear.